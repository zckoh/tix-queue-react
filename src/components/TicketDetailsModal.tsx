import { useState } from "react";
import moment from "moment";
import { Button, Modal, Form, Container, Col } from "react-bootstrap";
import { v4 as uuidv4 } from "uuid";

import useTicketForm from "../hooks/useFormHook";
import { postUpdateTicket } from "../services";
import { Status, Notification, Ticket, Department, Technician } from "../types";
import TicketDeleteModal from "./TicketDeleteModal";

import "../styles/Modal.css";

interface TicketDetailsModalPropTypes {
  ticket: Ticket | undefined;
  show: Boolean;
  allStatus: Array<Status>;
  allDepartments: Array<Department>;
  allTechnicians: Array<Technician>;
  handleClose: () => void;
  handleNotification: (notification: Notification) => void;
}

const TicketDetailsModal = (props: TicketDetailsModalPropTypes) => {
  const { show, ticket, allStatus, allDepartments, allTechnicians, handleClose, handleNotification } = props;

  const [amUpdating, setUpdating] = useState(false);
  const [isLoading, setLoading] = useState(false);
  const [isUpdated, setUpdated] = useState(false);
  const [isError, setError] = useState(false);

  const formatDateString = (date: string | undefined) => moment(date).format("DD/MM/YYYY HH:mm:ss");

  const toggleUpdate = () => {
    setUpdating(!amUpdating);
  };

  const updateButtonVariant = () => {
    if (isError) {
      return `danger`;
    }

    if (isUpdated) {
      return `success`;
    }

    return `primary`;
  };

  const updateTicket = () => {
    // Prepare JSON object
    const { title, author, description, createdAt, id, solution } = inputs;

    const ticketToUpdate = {
      id,
      title,
      author,
      description,
      status: allStatus?.find((status) => status.name === inputs.status),
      createdAt,
      solution: inputs.status !== "Completed" ? null : solution,
      department: allDepartments?.find((department) => department.name === inputs.department),
      technician: allTechnicians?.find((technician) => technician.name === inputs.technician),
    };

    setError(false);
    setLoading(true);

    // Post new ticket data
    postUpdateTicket(ticketToUpdate)
      .then(() => {
        setLoading(false);
        setUpdated(true);
        handleNotification({
          notificationId: uuidv4(),
          message: "✔️ - Ticket updated successfully.",
          status: "success",
          show: true,
          createdAt: moment(),
        });
        closeViewModal();
      })
      .catch(() => {
        setLoading(false);
        setError(true);
        handleNotification({
          notificationId: uuidv4(),
          message: "❌ - Failed to update ticket.",
          status: "failed",
          show: true,
          createdAt: moment(),
        });
      });
  };

  const { inputs, handleInputChange, handleSubmit, setInputs } = useTicketForm(
    {
      id: 0,
      title: "",
      author: "",
      description: "",
      status: "",
      createdAt: "",
      solution: "",
      department: "N/A",
      technician: "N/A",
    },
    updateTicket,
  );

  const onDepartmentChange = (event: any) => {
    if (inputs.department !== event.target.value) setInputs({ ...inputs, technician: "N/A" });
    handleInputChange(event);
  };

  const populateModal = () => {
    setInputs({
      id: ticket?.id,
      title: ticket?.title,
      author: ticket?.author,
      description: ticket?.description,
      status: ticket?.status?.name,
      createdAt: ticket?.createdAt,
      solution: ticket?.solution ? ticket.solution : "",
      department: ticket?.department ? ticket.department.name : "N/A",
      technician: ticket?.technician ? ticket.technician.name : "N/A",
    });
  };

  const closeViewModal = () => {
    setInputs({
      title: "",
      author: "",
      description: "",
      status: "",
      createdAt: "",
      solution: "",
      department: "N/A",
      technician: "N/A",
    });
    setUpdating(false);
    setUpdated(false);
    setLoading(false);
    setError(false);
    handleClose();
  };

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  const handleCloseDeleteModal = () => setShowDeleteModal(false);
  const handleShowDeletModal = () => setShowDeleteModal(true);

  return (
    <Modal size="lg" show={show} onShow={populateModal} onHide={closeViewModal} animation={true} className="modal-dark">
      <Form className="modal-form-input" onSubmit={handleSubmit}>
        <Container className="pb-3">
          <Modal.Header closeButton>
            <Modal.Title>{inputs?.title}</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <div className="d-flex justify-content-end">
              <Button className="mr-1" variant={amUpdating ? "secondary" : "success"} onClick={toggleUpdate}>
                {amUpdating ? `Disable Updating` : `Enable Updating`}
              </Button>
              <Button className="ml-1" variant="danger" onClick={handleShowDeletModal}>
                Delete
              </Button>
            </div>
            <Form.Group controlId="ticketTitle">
              <Form.Label>Title</Form.Label>
              <Form.Control
                required
                type="text"
                name="title"
                value={inputs?.title}
                disabled={!amUpdating}
                onChange={handleInputChange}
              />
            </Form.Group>
            <Form.Group controlId="ticketAuthor">
              <Form.Label>Author</Form.Label>
              <Form.Control
                required
                type="email"
                name="author"
                placeholder="Enter email address"
                value={inputs?.author}
                disabled={!amUpdating}
                onChange={handleInputChange}
              />
              <Form.Text className="text-muted">Please enter the email address of the ticket author.</Form.Text>
            </Form.Group>
            <Form.Row>
              <Form.Group as={Col} controlId="ticketDepartment">
                <Form.Label>Department</Form.Label>
                <Form.Control
                  required
                  as="select"
                  name="department"
                  value={inputs?.department}
                  disabled={!amUpdating}
                  onChange={onDepartmentChange}
                >
                  <option>N/A</option>
                  {allDepartments?.map((department, i) => (
                    <option key={i}>{department.name}</option>
                  ))}
                </Form.Control>
                <Form.Text className="text-muted">Department for the ticket to be assigned to.</Form.Text>
              </Form.Group>
              <Form.Group as={Col} controlId="ticketTechnician">
                <Form.Label>Assignee</Form.Label>
                <Form.Control
                  required
                  as="select"
                  name="technician"
                  value={inputs?.technician}
                  disabled={!amUpdating}
                  onChange={handleInputChange}
                >
                  <option>N/A</option>
                  {allTechnicians
                    ?.filter((technician) =>
                      inputs?.department === "N/A" ? true : technician.department?.name === inputs?.department,
                    )
                    ?.map((technician, i) => (
                      <option key={i}>{technician.name}</option>
                    ))}
                </Form.Control>
                <Form.Text className="text-muted">Assignee who will work on the ticket.</Form.Text>
              </Form.Group>
            </Form.Row>
            <Form.Row>
              <Form.Group as={Col} controlId="ticketStatus">
                <Form.Label>Status</Form.Label>
                <Form.Control
                  required
                  as="select"
                  name="status"
                  value={inputs?.status}
                  disabled={!amUpdating}
                  onChange={handleInputChange}
                >
                  {allStatus?.map((status, i) => (
                    <option key={i}>{status.name}</option>
                  ))}
                </Form.Control>
              </Form.Group>
              <Form.Group as={Col} controlId="ticketTimestamp">
                <Form.Label>Created At</Form.Label>
                <Form.Control
                  required
                  type="text"
                  name="createdAt"
                  defaultValue={formatDateString(ticket?.createdAt)}
                  disabled
                />
              </Form.Group>
            </Form.Row>
            <Form.Group controlId="ticketDescription">
              <Form.Label>Description</Form.Label>
              <Form.Control
                as="textarea"
                rows={3}
                name="description"
                placeholder="Enter ticket description"
                value={inputs?.description}
                disabled={!amUpdating}
                onChange={handleInputChange}
              />
              <Form.Text className="text-muted">
                This description will help the techinician to understand the issue you are facing.
              </Form.Text>
            </Form.Group>
            {inputs.status === "Completed" && (
              <Form.Group controlId="ticketSolution">
                <Form.Label>Solution</Form.Label>
                <Form.Control
                  as="textarea"
                  rows={3}
                  placeholder="Enter ticket solution"
                  name="solution"
                  disabled={!amUpdating}
                  onChange={handleInputChange}
                  value={inputs.solution}
                />
                <Form.Text className="text-muted">
                  This solution will help other users to solve similar issues.
                </Form.Text>
              </Form.Group>
            )}
          </Modal.Body>
          <Modal.Footer>
            <Button variant="secondary" onClick={closeViewModal}>
              {amUpdating ? `Cancel` : `Close`}
            </Button>
            <Button type="submit" active={isLoading} disabled={isUpdated} variant={updateButtonVariant()}>
              {isUpdated
                ? `Ticket Updated`
                : isLoading
                ? `Updating...`
                : isError
                ? `Failed to update ticket`
                : // Default
                  `Save Changes`}
            </Button>
          </Modal.Footer>
        </Container>
      </Form>

      <TicketDeleteModal
        show={showDeleteModal}
        handleClose={handleCloseDeleteModal}
        handleSuccessDelete={closeViewModal}
        ticketTitle={inputs.title}
        ticketId={inputs.id}
        handleNotification={handleNotification}
      />
    </Modal>
  );
};

export default TicketDetailsModal;
