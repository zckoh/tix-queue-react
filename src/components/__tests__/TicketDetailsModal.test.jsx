import { mount } from "enzyme";
import { waitFor } from "@testing-library/react";
import moment from "moment";
import { enableFetchMocks } from "jest-fetch-mock";
enableFetchMocks();
beforeEach(() => {
  fetch.resetMocks();
});

import { Button, Modal, Form } from "react-bootstrap";

import TicketDetailsModal from "../TicketDetailsModal";

const ticket = {
  id: 1,
  title: "Test Ticket 1",
  author: "author2@example.com",
  description: "This is a multiple line description\nMany words inside",
  createdAt: "2021-03-28T19:46:00.429912",
  status: {
    id: 2,
    name: "In Progress",
  },
};

const updatedTicket = {
  id: 1,
  title: "Test Ticket 1",
  author: "author2@example.com",
  description: "This is a updated ticket description",
  createdAt: "2021-03-28T19:46:00.429912",
  status: {
    id: 2,
    name: "In Progress",
  },
};

const allStatus = [
  {
    id: 1,
    name: "New",
  },
  {
    id: 2,
    name: "In Progress",
  },
  {
    id: 3,
    name: "Completed",
  },
];

const allTechnicians = [
  {
    id: 4,
    name: "Zac Koh",
    department: {
      id: 2,
      name: "IT",
    },
  },
  {
    id: 6,
    name: "Sam smith",
    department: {
      id: 3,
      name: "Cloud",
    },
  },
  {
    id: 7,
    name: "post.malone@helpqueue.com",
    department: {
      id: 3,
      name: "Cloud",
    },
  },
];

const allDepartments = [
  {
    id: 1,
    name: "Hardware",
  },
  {
    id: 2,
    name: "IT",
  },
  {
    id: 3,
    name: "Cloud",
  },
  {
    id: 9,
    name: "AWS",
  },
];
describe("<TicketDetailsModal />", () => {
  const mockHandleClose = jest.fn();
  const mockHandleNotification = jest.fn();
  const wrapper = mount(
    <TicketDetailsModal
      ticket={ticket}
      show={false}
      allStatus={allStatus}
      handleClose={mockHandleClose}
      handleNotification={mockHandleNotification}
      allTechnicians={allTechnicians}
      allDepartments={allDepartments}
    />,
  );

  it(`should correctly renders component`, () => {
    expect(wrapper).toMatchSnapshot();
  });

  it(`should display ticket information when modal is shown`, () => {
    // Show Modal
    wrapper.setProps({ show: true });
    wrapper.update();
    const ticketDetailsModal = wrapper.find(Modal).first();
    expect(ticketDetailsModal.props().show).toBeTruthy();

    const modalTitle = ticketDetailsModal.find(Modal.Title);
    expect(modalTitle.props().children).toEqual(ticket.title);

    const ticketTitle = ticketDetailsModal
      .find(Form.Control)
      .findWhere((n) => n.props().name === "title")
      .first();
    expect(ticketTitle.props().value).toEqual(ticket.title);

    const ticketAuthor = ticketDetailsModal
      .find(Form.Control)
      .findWhere((n) => n.props().name === "author")
      .first();
    expect(ticketAuthor.props().value).toEqual(ticket.author);

    const ticketDescription = ticketDetailsModal
      .find(Form.Control)
      .findWhere((n) => n.props().name === "description")
      .first();
    expect(ticketDescription.props().value).toEqual(ticket.description);

    const ticketStatus = ticketDetailsModal
      .find(Form.Control)
      .findWhere((n) => n.props().name === "status")
      .first();
    expect(ticketStatus.props().value).toEqual(ticket.status.name);

    const ticketCreatedAt = ticketDetailsModal
      .find(Form.Control)
      .findWhere((n) => n.props().name === "createdAt")
      .first();
    expect(ticketCreatedAt.props().defaultValue).toEqual(moment(ticket.createdAt).format("DD/MM/YYYY HH:mm:ss"));
  });

  describe(`Update Toggle Button`, () => {
    it(`should enable editing on Form.Control once clicked`, () => {
      const updateButton = wrapper.find(Button).first();
      expect(updateButton.props().variant).toEqual(`success`);

      updateButton.simulate("click");
      wrapper.update();

      const ticketTitle = wrapper
        .find(Modal)
        .first()
        .find(Form.Control)
        .findWhere((n) => n.props().name === "title")
        .first();
      expect(ticketTitle.props().disabled).toBeFalsy();
    });
  });

  describe(`Updating Ticket`, () => {
    it(`successful update should update button to show updated message `, async () => {
      fetch.mockResponseOnce(JSON.stringify(updatedTicket), { status: 202 });
      const ticketDescription = wrapper.find("textarea").first();
      ticketDescription.instance().value = "foo";

      const updateButton = wrapper
        .find(Button)
        .findWhere((n) => n.props().type === "submit")
        .first();
      updateButton.simulate("submit");

      await waitFor(() => {
        expect(mockHandleClose).toHaveBeenCalledTimes(1);
        expect(mockHandleNotification).toHaveBeenCalledTimes(1);
      });
    });
  });
});
