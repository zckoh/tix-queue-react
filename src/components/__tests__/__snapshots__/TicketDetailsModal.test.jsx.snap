// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketDetailsModal /> should correctly renders component 1`] = `ReactWrapper {}`;
